manifest/manifestlist/manifestlist.go

```diff
@@ -123,6 +123,16 @@ type ManifestList struct {
 // References returns the distribution descriptors for the referenced image
 // manifests.
 func (m ManifestList) References() []distribution.Descriptor {
+	return m.ManifestReferences()
+}
+
+// BlobReferences returns blob descriptors referenced by this manifest.
+func (sm ManifestList) BlobReferences() []distribution.Descriptor {
+	return nil
```

Comment:
Part of the reason I was interested in this codebase and looking at how it handled more generic dependency graphs was because I have an interest in having my own custom manifest type that has more complex dependencies. So it's definitely my preference this is written in a generic way.

Interesting point about some other types of manifests having manifest dependencies. I agree it does look like e.g. schema2 can have references to schema1 or schema2 manifets according to the implementation of verifyManifest. However I don't see anything that explains that in https://docs.docker.com/registry/spec/manifest-v2-2/ and I'm not sure what it would mean. Maybe it's just a copy-paste from the manifest list implementation?

Reply:
> Maybe it's just a copy-paste from the manifest list implementation?

Gosh, I would hope not :sweat: 

> Part of the reason I was interested in this codebase and looking at how it handled more generic dependency graphs was because I have an interest in having my own custom manifest type that has more complex dependencies. So it's definitely my preference this is written in a generic way.

@msg555 gotcha, I'm hesitant to include those changes along with this MR. Downstream projects use us a library and making changes to these types here needs to be preceded by an in-depth, dedicated discussion and probably shouldn't be included without a major version bump.

My feelings on this is that we should split this work up by:

1. keeping the garbage collection related changes in this MR, using type casting to discover the manifest lists, and updating the title to fit this, as the GC removing manifests removed by a manifest list is a serious bug
2. You opening an issue to discuss the generic dependency graphs and discuss your use case and implementation
3. Me opening an issue to investigate the mysterious validation code 

WDYT?